File: api/_lib/content-autopilot/higgsfield.ts
import {
  HIGGSFIELD_API_BASE,
  HIGGSFIELD_IMAGE_MODEL,
  HIGGSFIELD_IMAGE_RESOLUTION,
  isSuccessfulHiggsfieldStatus,
  isTerminalHiggsfieldStatus,
} from "../../../supabase/functions/_shared/higgsfieldImage.ts";

export const KLING_I2V_PATH = "/kling-video/v3.0/pro/image-to-video";
export const KLING_I2V_ESTIMATE_PATH = "/estimate/kling-video/v3.0/pro/image-to-video";
export const KLING_DURATION_SECONDS = 5;

function readHiggsfieldAuthorizationFromProcessEnv(): string | null {
  const combined = String(process.env.HF_CREDENTIALS || process.env.HF_KEY || "").trim();
  const splitAt = combined.indexOf(":");
  if (splitAt > 0) {
    const keyId = combined.slice(0, splitAt).trim();
    const secret = combined.slice(splitAt + 1).trim();
    if (keyId && secret) return `Key ${keyId}:${secret}`;
  }
  const keyId = String(process.env.HF_API_KEY_ID || process.env.HF_API_KEY || "").trim();
  const secret = String(
    process.env.HF_API_KEY_SECRET || process.env.HF_SECRET || process.env.HF_API_SECRET || "",
  ).trim();
  if (keyId && secret) return `Key ${keyId}:${secret}`;
  return null;
}

export function higgsfieldConfigured(): boolean {
  return readHiggsfieldAuthorizationFromProcessEnv() !== null;
}

async function higgsfieldJson(
  authorization: string,
  method: string,
  path: string,
  body?: unknown,
): Promise<Record<string, unknown>> {
  const url = path.startsWith("http") ? path : `${HIGGSFIELD_API_BASE}${path}`;
  const response = await fetch(url, {
    method,
    headers: {
      Authorization: authorization,
      Accept: "application/json",
      "Content-Type": "application/json",
      "User-Agent": "tdg-content-autopilot/higgsfield",
    },
    body: body === undefined ? undefined : JSON.stringify(body),
  });
  const text = await response.text();
  let parsed: Record<string, unknown> = {};
  if (text) {
    try {
      parsed = JSON.parse(text) as Record<string, unknown>;
    } catch {
      parsed = {};
    }
  }
  if (!response.ok) {
    throw new Error(`higgsfield_http_${response.status}`);
  }
  return parsed;
}

export async function uploadBytesToHiggsfield(
  authorization: string,
  bytes: Uint8Array,
  contentType: string,
): Promise<string> {
  const meta = await higgsfieldJson(authorization, "POST", "/files/generate-upload-url", {
    content_type: contentType,
  });
  const uploadUrl = String(meta.upload_url || "");
  const publicUrl = String(meta.public_url || "");
  if (!uploadUrl || !publicUrl) throw new Error("higgsfield_upload_meta_missing");
  const uploadHeaders = new Headers();
  const rawHeaders = meta.upload_headers;
  if (rawHeaders && typeof rawHeaders === "object") {
    for (const [key, value] of Object.entries(rawHeaders as Record<string, unknown>)) {
      if (typeof value === "string") uploadHeaders.set(key, value);
    }
  }
  if (!uploadHeaders.has("Content-Type")) uploadHeaders.set("Content-Type", contentType);
  const putBody = bytes.buffer.slice(bytes.byteOffset, bytes.byteOffset + bytes.byteLength) as ArrayBuffer;
  const uploaded = await fetch(uploadUrl, { method: "PUT", headers: uploadHeaders, body: putBody });
  if (!uploaded.ok) throw new Error("higgsfield_upload_failed");
  return publicUrl;
}

export async function submitContentImage(
  authorization: string,
  prompt: string,
): Promise<string> {
  const submitted = await higgsfieldJson(authorization, "POST", `/${HIGGSFIELD_IMAGE_MODEL}`, {
    prompt,
    aspect_ratio: "9:16",
    resolution: HIGGSFIELD_IMAGE_RESOLUTION,
  });
  const requestId = String(submitted.request_id || "").trim();
  if (!requestId) throw new Error("higgsfield_image_request_missing");
  return requestId;
}

export function extractHiggsfieldImageUrl(body: Record<string, unknown>): string | null {
  const images = body.images;
  if (Array.isArray(images)) {
    for (const image of images) {
      if (image && typeof image === "object" && typeof (image as { url?: unknown }).url === "string") {
        return (image as { url: string }).url;
      }
      if (typeof image === "string" && /^https?:\/\//i.test(image)) return image;
    }
  }
  if (typeof body.image_url === "string") return body.image_url;
  return null;
}

export function extractHiggsfieldVideoUrl(body: Record<string, unknown>): string | null {
  const video = body.video;
  if (video && typeof video === "object" && typeof (video as { url?: unknown }).url === "string") {
    return (video as { url: string }).url;
  }
  if (typeof body.video_url === "string") return body.video_url;
  return null;
}

export async function pollHiggsfieldRequest(
  authorization: string,
  requestId: string,
  options?: { maxAttempts?: number; sleepMs?: number },
): Promise<{ done: boolean; failed: boolean; body: Record<string, unknown> | null }> {
  const maxAttempts = options?.maxAttempts ?? 60;
  const sleepMs = options?.sleepMs ?? 3000;
  for (let attempt = 0; attempt < maxAttempts; attempt += 1) {
    const current = await higgsfieldJson(authorization, "GET", `/requests/${requestId}/status`);
    const status = String(current.status || "");
    if (isTerminalHiggsfieldStatus(status)) {
      return {
        done: true,
        failed: !isSuccessfulHiggsfieldStatus(status),
        body: current,
      };
    }
    if (attempt < maxAttempts - 1) {
      await new Promise((resolve) => setTimeout(resolve, sleepMs));
    }
  }
  return { done: false, failed: false, body: null };
}

export async function estimateKlingVideoUsd(authorization: string, imageUrl: string, prompt: string): Promise<number> {
  const estimate = await higgsfieldJson(authorization, "POST", KLING_I2V_ESTIMATE_PATH, {
    image_url: imageUrl,
    prompt,
    duration: KLING_DURATION_SECONDS,
    sound: "off",
  });
  const usd = Number(estimate.usd);
  if (!Number.isFinite(usd) || usd < 0) throw new Error("higgsfield_estimate_missing");
  return usd;
}

export async function submitKlingVideo(
  authorization: string,
  imageUrl: string,
  prompt: string,
): Promise<string> {
  const submitted = await higgsfieldJson(authorization, "POST", KLING_I2V_PATH, {
    image_url: imageUrl,
    prompt,
    duration: KLING_DURATION_SECONDS,
    sound: "off",
  });
  const requestId = String(submitted.request_id || "").trim();
  if (!requestId) throw new Error("higgsfield_video_request_missing");
  return requestId;
}

export function getHiggsfieldAuthorization(): string {
  const auth = readHiggsfieldAuthorizationFromProcessEnv();
  if (!auth) throw new Error("higgsfield_credentials_missing");
  return auth;
}

File: scripts/reel-director/higgsfield.py
"""Higgsfield client reused from scripts/generate_lauren_overwhelm_reel.py.

Estimate and upload are not video generation. Submit is paid and must stay gated.
"""

from __future__ import annotations

import json
import os
import time
import urllib.error
import urllib.request
from pathlib import Path

HIGGSFIELD_API_BASE = "https://api.higgsfield.ai"
MODEL_ID = "kling-video/v3.0/pro/image-to-video"
ESTIMATE_PATH = "/estimate/kling-video/v3.0/pro/image-to-video"
SUBMIT_PATH = "/kling-video/v3.0/pro/image-to-video"
UA = "tdg-reel-director/higgsfield"


def read_credentials() -> tuple[str, str]:
    combined = (os.environ.get("HF_CREDENTIALS") or os.environ.get("HF_KEY") or "").strip()
    if ":" in combined:
        key_id, secret = combined.split(":", 1)
        key_id, secret = key_id.strip(), secret.strip()
        if key_id and secret:
            return key_id, secret
    key_id = (os.environ.get("HF_API_KEY_ID") or os.environ.get("HF_API_KEY") or "").strip()
    secret = (
        os.environ.get("HF_API_KEY_SECRET") or os.environ.get("HF_SECRET") or os.environ.get("HF_API_SECRET") or ""
    ).strip()
    if key_id and secret:
        return key_id, secret
    raise SystemExit("BLOCKED: Higgsfield credentials missing. No Replicate fallback.")


def credentials_present() -> bool:
    try:
        read_credentials()
        return True
    except SystemExit:
        return False


def auth_header() -> str:
    key_id, secret = read_credentials()
    return f"Key {key_id}:{secret}"


def hf_request(method: str, path_or_url: str, data: dict | None = None, timeout: int = 180) -> dict:
    url = path_or_url if path_or_url.startswith("http") else f"{HIGGSFIELD_API_BASE}{path_or_url}"
    body = None if data is None else json.dumps(data).encode()
    headers = {
        "Authorization": auth_header(),
        "User-Agent": UA,
        "Accept": "application/json",
    }
    if body is not None:
        headers["Content-Type"] = "application/json"
    req = urllib.request.Request(url, data=body, method=method, headers=headers)
    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            raw = resp.read().decode()
            return json.loads(raw) if raw else {}
    except urllib.error.HTTPError as e:
        err = e.read().decode()
        raise RuntimeError(f"{method} {url} -> {e.code} {err[:800]}") from e


def parse_estimate(body: object) -> dict:
    if not isinstance(body, dict):
        raise RuntimeError(f"estimate response not an object: {body!r}")
    usd_raw = body.get("usd")
    usd = float(usd_raw) if usd_raw is not None else float("nan")
    if not (usd >= 0):
        raise RuntimeError(f"live estimate missing usd: {body}")
    credits = body.get("credits")
    return {"usd": usd, "credits": None if credits is None else str(credits), "raw": body}


def video_url_from_status(body: dict) -> str | None:
    video = body.get("video")
    if isinstance(video, dict) and isinstance(video.get("url"), str):
        return video["url"]
    if isinstance(body.get("video_url"), str):
        return body["video_url"]
    images = body.get("images")
    if isinstance(images, list) and images:
        first = images[0]
        if isinstance(first, dict) and isinstance(first.get("url"), str):
            return first["url"]
    return None


def clip_payload(image_url: str, prompt: str, duration: int) -> dict:
    return {
        "image_url": image_url,
        "prompt": prompt,
        "duration": int(duration),
        "sound": "off",
    }


def upload_image(path: Path) -> str:
    ctype = "image/png" if path.suffix.lower() == ".png" else "image/jpeg"
    meta = hf_request("POST", "/files/generate-upload-url", {"content_type": ctype})
    public_url = meta.get("public_url")
    upload_url = meta.get("upload_url")
    if not public_url or not upload_url:
        raise RuntimeError(f"upload URL missing fields: keys={list(meta.keys())}")
    headers = dict(meta.get("upload_headers") or {"Content-Type": ctype})
    req = urllib.request.Request(str(upload_url), data=path.read_bytes(), method="PUT", headers=headers)
    with urllib.request.urlopen(req, timeout=180) as resp:
        resp.read()
    return str(public_url)


def verify_auth() -> None:
    meta = hf_request("POST", "/files/generate-upload-url", {"content_type": "image/png"})
    if not meta.get("upload_url"):
        raise RuntimeError("HIGGSFIELD_AUTH FAIL: upload URL response missing upload_url")


def poll_job(request_id: str, timeout_s: int = 1200) -> dict:
    start = time.time()
    while time.time() - start < timeout_s:
        current = hf_request("GET", f"/requests/{request_id}/status")
        status = str(current.get("status") or "")
        print(f"  {request_id} {status}", flush=True)
        if status.lower() in {"completed", "succeeded", "failed", "nsfw", "canceled", "cancelled", "error"}:
            return current
        time.sleep(8)
    raise TimeoutError(f"Higgsfield job {request_id} timed out")


def download(url: str, dest: Path) -> None:
    dest.parent.mkdir(parents=True, exist_ok=True)
    req = urllib.request.Request(url, headers={"User-Agent": UA})
    with urllib.request.urlopen(req, timeout=180) as resp, dest.open("wb") as f:
        while True:
            chunk = resp.read(1024 * 256)
            if not chunk:
                break
            f.write(chunk)
